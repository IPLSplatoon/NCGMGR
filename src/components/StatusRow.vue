<template>
    <div class="status-row layout horizontal center-vertical">
        <div
            class="status-row__status-light"
            :class="`color-${color}`"
        />
        <slot />
    </div>
</template>

<script lang="ts">
import { defineComponent, PropType } from 'vue'

export type StatusRowColor = 'red' | 'yellow' | 'green' | 'gray'

export default defineComponent({
    name: 'StatusRow',

    props: {
        color: {
            type: String as PropType<StatusRowColor>,
            required: true,
            validator: (value: string) => {
                return ['red', 'yellow', 'green', 'gray'].includes(value)
            }
        }
    }
})
</script>

<style lang="scss" scoped>
@import 'src/styles/colors';

@mixin status-light-color($color) {
    background-color: $color;
    filter: drop-shadow(0 0 3px $color);
}

.status-row {
    overflow-wrap: anywhere;

    .status-row__status-light {
        min-width: 12px;
        height: 12px;
        border-radius: 50%;
        margin: 0 8px 0 4px;

        &.color-red {
            @include status-light-color($red)
        }

        &.color-yellow {
            @include status-light-color($yellow)
        }

        &.color-green {
            @include status-light-color($green)
        }

        &.color-gray {
            @include status-light-color($input-color)
        }
    }
}
</style>
