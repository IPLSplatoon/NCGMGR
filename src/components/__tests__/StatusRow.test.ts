import StatusRow from '@/components/StatusRow.vue'
import { shallowMount } from '@vue/test-utils'

describe('StatusRow', () => {
    it('matches snapshot', () => {
        const wrapper = shallowMount(StatusRow, {
            props: {
                color: 'yellow'
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
    })

    it('applies color class to status light', () => {
        const wrapper = shallowMount(StatusRow, {
            props: {
                color: 'red'
            }
        })

        expect(wrapper.get('.status-row__status-light').classes()).toContain('color-red')
    })

    describe('validator: color', () => {
        const validator = StatusRow.props.color.validator

        it('returns true for known colors', () => {
            expect(validator('red')).toEqual(true)
            expect(validator('yellow')).toEqual(true)
            expect(validator('green')).toEqual(true)
            expect(validator('gray')).toEqual(true)
        })

        it('returns false for unknown colors', () => {
            expect(validator('pink')).toEqual(false)
            expect(validator(undefined)).toEqual(false)
            expect(validator(null)).toEqual(false)
            expect(validator('')).toEqual(false)
        })
    })
})
