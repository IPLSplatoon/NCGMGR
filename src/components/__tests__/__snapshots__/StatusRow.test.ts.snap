// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusRow matches snapshot 1`] = `
"<div class="status-row layout horizontal center-vertical">
  <div class="status-row__status-light color-yellow"></div>
</div>"
`;
